File: lib/hash_tree.py
import hashlib
import math
import copy

CHUNK_SIZE = 1

def hash(x):
    m = hashlib.sha256()
    m.update(x)
    return m.digest()


def hash_bottom(input):
    chunk_hashes = [hash(input.read(CHUNK_SIZE))]
    while True:
        chunk = input.read(CHUNK_SIZE)
        if not chunk:
            break
        chunk_hashes.append(hash(chunk))

    return chunk_hashes

def hash_layer(lst):
    chunk_hashes = []
    lst = copy.copy(lst)
    while lst:
        a = lst.pop(0)
        try:
            b = lst.pop(0)
            chunk_hashes.append(hash(a + b))
        except IndexError:
            chunk_hashes.append(a)
    return chunk_hashes

def hash_tree(input):
    levels = [hash_bottom(input)]
    while len(levels[-1]) > 1:
        levels.append(hash_layer(levels[-1]))
    for layer in reversed(levels):
        for hash in layer:
            yield hash

File: js/hash_tree.js
;(function(sha256, global, undefined){
  var HashTree = global.HashTree = function(input)
  {
    var levels = [hash_bottom(input)]
      , last_level = levels[levels.length - 1];

    while (last_level.length > 1)
    {
      levels.push(hash_layer(last_level));

      last_level = levels[levels.length - 1];
    }

    return {
      levels: levels,
      root_hash: levels[levels.length -1][0]
    };
  };
  HashTree.CHUNK_SIZE = 1024;

  var hash = HashTree.hash = function (input) {
    return sha256(input, {asBytes: true});
  };

  var hash_bottom = HashTree.hash_bottom = function(input) {
    var i = HashTree.CHUNK_SIZE
      , chunk_hashes = [hash(input.slice(0, HashTree.CHUNK_SIZE))]
      , chunk;
    
    while (chunk = input.slice(i, i + HashTree.CHUNK_SIZE))
    {
      chunk_hashes.push(hash(chunk));
      i += HashTree.CHUNK_SIZE;
    }

    return chunk_hashes;
  };

  var hash_layer = HashTree.hash_layer = function(lst) {
    var chunk_hashes = []
      , a
      , b
      , lst = lst.slice();

    while ( a = lst.shift() )
    {
      if ( b = lst.shift() )
        chunk_hashes.push(hash(a.concat(b)));
      else
        chunk_hashes.push(a);
    }
      
    return chunk_hashes;
  };


})(Crypto.SHA256, this);
